tests/mbt/model/ccv_model.qnt

```diff
@@ -632,4 +654,270 @@ module ccv_model {
+            nondet node = oneOf(nodes)
+            nondet consumerAddr = oneOf(consumerAddresses)
+            nondet consumer = oneOf(runningConsumers)
+            KeyAssignment(consumer, node, consumerAddr),
+        }
+
+    action KeyAssignment(
+        chain: Chain,
+        validator: Node,
+        consumerAddr: ConsumerAddr
+    ): bool =
+        val result = assignConsumerKey(currentState, chain, validator, consumerAddr)
+        all {
+            hasError(result) == false,
+            currentState' = result.newState,
+            trace' = trace.append(
+                {...emptyAction,
+                kind: "KeyAssignment",
+                consumerChain: chain,
+                validator: validator,
+                consumerAddr: consumerAddr
+                }
+            ),
+            params' = params,
+        }
+
+    // invariants for key assignment - some invariants are in addition, some need to be adjusted from the original model
+
+    // Every validator set on any consumer chain MUST either be or have been 
+    // a validator set on the provider chain, under the key assignment at the time.
+    val providerKeyAssignedValSetHistory = currentState.providerState.keyAssignedValSetHistory
+
+    val ValidatorSetHasExistedKeyAssignmentInv =
+        runningConsumers.forall(chain =>
+            currentState.consumerStates.get(chain).chainState.votingPowerHistory.toSet().forall(
+                validatorSet => providerKeyAssignedValSetHistory.getOrElse(chain, List()).toSet().exists(
+                    provValSet => removeZeroPowers(provValSet) == removeZeroPowers(validatorSet)
+                )
+            )
+        )
+
+    // Any update in the power of a validator on the provider
+    // MUST be present in a ValidatorSetChangePacket that is sent to all registered consumer chains,
+    // and the key assignment of each validator should be applied in that VSCPacket.
+    val ValidatorUpdatesArePropagatedKeyAssignmentInv = 
+        // when the provider has just entered a validator set into a block...
+        ValUpdatePrecondition and currentState.providerState.providerValidatorSetChangedInThisBlock
+            implies
+        val providerValSetInCurBlock = providerValidatorHistory.head()
+        // ... for each consumer that is running then ...
+        runningConsumers.forall(
+            // ...the validator set under key assignment is in a sent packet...
+            val providerState = currentState.providerState
+            consumer => providerState.sentVscPacketsToConsumer.get(consumer).toSet().exists(
+                packet => 
+                    packet.validatorSet == 
+                    applyKeyAssignmentToValSet(providerState, consumer, providerValSetInCurBlock)
+            )
+            // or the consumer was just started, which we detect by the consumer having a timestamp of 0
+            // and the consumer having the validator set that was just sent in the block
+            or 
```

Comment:
> currentState.consumerStates.get(consumer).chainState.currentValidatorSet == providerValSetInCurBlock

Shouldn't this also have the key assignments applied?

Reply:
General feedback on the invariants: Some of them are very hard to read as they are written in a very concise form. Maybe it's just a matter of getting used to quint though.

Reply:
I'll add some more comments to the invariants, specifically.
It's why I also try to split them up sometimes, but some are just not easily decomposed. I'll see if I can split them more

Reply:
Also seems I was wrong here - the precondition is fine, rereading the property, it only talks about updates to validator powers, not to changes in key assignment.
The problem is that the or statement never mattered, since it's already an invariant with just the first disjunct, so the solution was to remove the second disjunct :)